### PR TITLE
Archive components for blog

### PR DESCRIPTION
**Pull request by edhardie**
@LukeTowers 

This looks to be a lot tidier! Sorry about the last one, still figuring out what happened.

### CODE REVIEW

File: Plugin.php [hunk 1]
```diff
@@ -27,7 +27,9 @@ public function registerComponents()
             'RainLab\Blog\Components\Post'       => 'blogPost',
             'RainLab\Blog\Components\Posts'      => 'blogPosts',
             'RainLab\Blog\Components\Categories' => 'blogCategories',
-            'RainLab\Blog\Components\RssFeed'    => 'blogRssFeed'
+            'RainLab\Blog\Components\RssFeed'    => 'blogRssFeed',
+	    'RainLab\Blog\Components\ArchiveList'  => 'blogArchiveList',
```
Comment (LukeTowers) [inline Plugin.php]: Missing space here

Comment (LukeTowers) [reply Plugin.php]: and there's an extra space in between `'RainLab\Blog\Components\ArchiveList'` and the `=>`

File: components/ArchiveList.php [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        $this->archiveMonths = $this->page['archiveMonths'] = $this->loadMonths($this->property('monthsToShow'));
+        $this->currentArchiveSlug = $this->page['currentArchiveSlug'] = $this->property('slug');
+        $this->archivePage = $this->page['archivePage'] = $this->property('archivePage');
+    }
+
+    protected function loadMonths($monthsToShow)
+    {
+        if (($monthsToShow > 0) ? $limit_string = "LIMIT 0, $monthsToShow" : $limit_string = "");
+
+        $monthsArray = DB::select("SELECT DATE_FORMAT(published_at, '%M %Y') AS month FROM `rainlab_blog_posts` WHERE published=1 GROUP BY DATE_FORMAT(published_at, '%M %Y') ORDER BY published_at ASC $limit_string");
```
Comment (LukeTowers) [inline components/ArchiveList.php]: Does this query have to be a DB raw statement? Is there a way that you can accomplish this with an Eloquent query instead? Also, you should be using prepared statements and if you have to use a raw query string I'd like you to get the DB name from the model instead of hardcoding it in.

**Comment by edhardie** [comment]
Wasn't able to drop the raw statements completely but managed to get it down to just selects leveraging eloquent for everything else. I could go further and bring the logic out of the query and into PHP but don't see any need really. 

Cleaned up that function a fair bit, the counting stuff is way less DB intense now, so cheers for pointing these out.

**Comment by edhardie** [comment]
@LukeTowers forgot to ping you about this!

File: Plugin.php [hunk 3]
```diff
@@ -27,7 +27,9 @@ public function registerComponents()
             'RainLab\Blog\Components\Post'       => 'blogPost',
             'RainLab\Blog\Components\Posts'      => 'blogPosts',
             'RainLab\Blog\Components\Categories' => 'blogCategories',
-            'RainLab\Blog\Components\RssFeed'    => 'blogRssFeed'
+            'RainLab\Blog\Components\RssFeed'    => 'blogRssFeed',
+	        'RainLab\Blog\Components\ArchiveList' => 'blogArchiveList',
```
Comment (LukeTowers) [inline Plugin.php]: Indentation is screwed up here.

File: components/ArchiveList.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        if (($monthsToShow > 0) ? $limit = $monthsToShow : $limit = 100);
+
+        $archiveRange = BlogPost::select(DB::raw("DATE_FORMAT(published_at, '%M %Y') AS month, count(*) as count"))
+                                ->where('published', 1)
+                                ->orderBy('published_at', 'ASC')
+                                ->limit($limit)
+                                ->groupBy(DB::raw("DATE_FORMAT(published_at, '%M %Y')"))
+                                ->get();
+
+        foreach ($archiveRange as $post) {
```
Comment (LukeTowers) [inline components/ArchiveList.php]: This variable should be named `$month`, not `$post`.

File: components/ArchiveList.php [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            $post->url = $this->controller->pageUrl($this->property('archivePage'), ['slug' => urlencode($post->month)]);
+            
+            if (!$this->property('showPostCount')) {
```
Comment (LukeTowers) [inline components/ArchiveList.php]: Why is this required? If you want to toggle the display of the count of posts in a month, then shouldn't that be done in the rendering, not the data structure?

File: components/ArchivePosts.php [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    public function defineProperties()
+    {
+        return [
+            'slug' => [
+                'title'       => 'rainlab.blog::lang.settings.post_slug',
+                'description' => 'rainlab.blog::lang.settings.post_slug_description',
+                'default'     => '{{ :slug }}',
+                'type'        => 'string'
+            ],
+            'pageNumber' => [
```
Comment (LukeTowers) [inline components/ArchivePosts.php]: Please name this `currentPage` instead.

File: components/ArchivePosts.php [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        return [
+            'name'        => 'rainlab.blog::lang.settings.archive_posts_title',
+            'description' => 'rainlab.blog::lang.settings.archive_posts_description'
+        ];
+    }
+
+    public function defineProperties()
+    {
+        return [
+            'slug' => [
```
Comment (LukeTowers) [inline components/ArchivePosts.php]: Please name this `month` instead.

File: components/ArchiveList.php [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    public function onRun()
+    {
+        $this->archiveMonths = $this->page['archiveMonths'] = $this->loadMonths($this->property('monthsToShow'));
+        $this->currentArchiveSlug = $this->page['currentArchiveSlug'] = $this->property('slug');
+        $this->archivePage = $this->page['archivePage'] = $this->property('archivePage');
+    }
+
+    protected function loadMonths($monthsToShow)
+    {
+        if (($monthsToShow > 0) ? $limit = $monthsToShow : $limit = 100);
```
Comment (LukeTowers) [inline components/ArchiveList.php]: Why is the default limit `100`? That seems arbitrary?

File: components/ArchiveList.php [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+        $archiveRange = BlogPost::select(DB::raw("DATE_FORMAT(published_at, '%M %Y') AS month, count(*) as count"))
```
Comment (LukeTowers) [inline components/ArchiveList.php]: Could you explain more about why you're only loading the months and post counts in this component instead of loading the relevant posts within each month? Not against it, just would like more information on the decision :)

File: components/ArchiveList.php [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+        foreach ($archiveRange as $post) {
+            $post->url = $this->controller->pageUrl($this->property('archivePage'), ['slug' => urlencode($post->month)]);
+            
+            if (!$this->property('showPostCount')) {
+                $post->count = false;
+            }
+        }
+
+        return collect($archiveRange);
```
Comment (LukeTowers) [inline components/ArchiveList.php]: Is `$archiveRange` not already a collection instance at this point in time?

File: components/ArchivePosts.php [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+        $this->noPostsMessage = $this->page['noPostsMessage'] = $this->property('noPostsMessage');
+
+        /*
+         * Page links
+         */
+        $this->postPage = $this->page['postPage'] = $this->property('postPage');
+        $this->categoryPage = $this->page['categoryPage'] = $this->property('categoryPage');
+    }
+
+    protected function listPosts($slug)
```
Comment (LukeTowers) [inline components/ArchivePosts.php]: `$slug` should be `$month` here too

File: components/ArchivePosts.php [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+    {
+        try {
+            $slug = str_replace('+', ' ', $slug);
```
Comment (LukeTowers) [inline components/ArchivePosts.php]: What's the rationale behind using `str_replace` instead of `urldecode`? http://php.net/manual/en/function.urldecode.php?

File: components/ArchivePosts.php [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+         */
+        $posts = BlogPost::with('categories')->listFrontEnd([
+            'page'       => $this->property('pageNumber'),
+            'sort'       => $this->property('sortOrder'),
+            'perPage'    => $this->property('postsPerPage'),
+            'dateMin'    => $dateMin,
+            'dateMax'    => $dateMax
+        ]);
+
+        // dd($posts);
```
Comment (LukeTowers) [inline components/ArchivePosts.php]: debug statement, please remove

File: components/ArchivePosts.php [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+
+            $post->categories->each(function($category){
+                $category->setUrl($this->categoryPage, $this->controller);
+            });
+        });
+
+        return $posts;
+    }
+
+    protected function loadCategory()
```
Comment (LukeTowers) [inline components/ArchivePosts.php]: Is this left over from copying from the existing components?

File: components/archivelist/default.htm [hunk 15]
```diff
@@ -0,0 +1,7 @@
+{% set archiveMonths = __SELF__.archiveMonths %}
+
+<ul>
+{% for archiveMonth in archiveMonths %}
```
Comment (LukeTowers) [inline components/archivelist/default.htm]: Minor quibble, but could you do `{% for month in archiveMonths %}` instead?

File: lang/en/lang.php [hunk 16]
```diff
@@ -119,6 +119,17 @@
         'rssfeed_blog' => 'Blog page',
         'rssfeed_blog_description' => 'Name of the main blog page file for generating links. This property is used by the default component partial.',
         'rssfeed_title' => 'RSS Feed',
-        'rssfeed_description' => 'Generates an RSS feed containing posts from the blog.'
+        'rssfeed_description' => 'Generates an RSS feed containing posts from the blog.',
+	'archive_list_name' => 'Archive List',
```
Comment (LukeTowers) [inline lang/en/lang.php]: indentation

**Comment by LukeTowers** [comment]
@edhardie you'll have to resubmit again, make your changes on a dedicated branch just for this PR and then don't merge anything else into it please.

**Comment by edhardie** [comment]
@LukeTowers cleaned up the diff for you - let me know if any further changes are required.

File: components/ArchivePosts.php [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+        $this->prepareVars();
+        $this->posts = $this->page['posts'] = $this->listPosts($this->property('month'));
+
+        /*
+         * If the page number is not valid, redirect
+         */
+        if ($currentPageParam = $this->paramName('currentPage')) {
+            $currentPage = $this->property('currentPage');
+
+            if ($currentPage > ($lastPage = $this->posts->lastPage()) && $currentPage > 1)
```
Comment (LukeTowers) [inline components/ArchivePosts.php]: Please replace with the `{}` wrapper, we try to avoid the one line conditionals in October

File: components/ArchivePosts.php [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+         * Page links
+         */
+        $this->postPage = $this->page['postPage'] = $this->property('postPage');
+        $this->categoryPage = $this->page['categoryPage'] = $this->property('categoryPage');
+    }
+
+    protected function listPosts($month)
+    {
+        try {
+            $month = str_replace('+', ' ', $month);
```
Comment (LukeTowers) [inline components/ArchivePosts.php]: Where would the `+` come from?

**Comment by LukeTowers** [comment]
@edhardie looks good! Just address the two comments I made and we can get this merged in!

**Comment by edhardie** [comment]
@LukeTowers cool, sorted, fingers crossed that's good to go

File: components/ArchivePosts.php (hunk 18 above)
Comment (edhardie) [reply components/ArchivePosts.php]: It's from a urlencode that we do when we generate the frontend links, have updated to a more sensible urldecode

**Comment by LukeTowers** [comment]
@edhardie looks good, sent you a PM on slack for testing this further

**Comment by edhardie** [comment]
@LukeTowers rewritten the loadPosts function to not rely on DATE_FORMAT - not 100% happy because it requires much more database interaction but can't think of a better solution that's going to be supported by all the database drivers. Let me know if you have any suggestions 😃

**Comment by LukeTowers** [comment]
@edhardie https://6670ab80111b.octodock.com:8443/blog-archive/October+2017

none of your imported posts are published, but they show up in the archive list (https://6670ab80111b.octodock.com:8443/blog) but don’t show up on those pages. Also, the count seems to be off, or is there really just one post per month?

**Comment by bennothommo** [comment]
@edhardie Are you still interested in having this merged into the Blog plugin? If so, would you mind resolving the conflicts, and I will be happy to give it another test and merge it if it all works.

**Comment by bennothommo** [comment]
Closing as it has been over a month since any activity on this occurred and we are trying to figure out what issues are still relevant. If this is still something that you would like to see through to fruition please respond and we can get the ball rolling.
